thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp

```diff
@@ -114,37 +128,117 @@ class CJoinGroup : public CSimpleInterfaceOf<IInterface>
+        Owned<IFileIO> iFileIO;
+        Owned<IFileIOStream> iFileIOStream;
+        Linked<IOutputRowSerializer> serializer;
+        unsigned nesting = 0;
+        MemoryBuffer nested;
+
+    // IRowSerializerTarget impl.
+        virtual void put(size32_t len, const void * ptr) override
         {
-            if (rhsState.arr->rows)
+            if (nesting)
+                nested.append(len, ptr);
+            else
+                iFileIOStream->write(len, ptr);
+        }
+        virtual size32_t beginNested(size32_t count) override
+        {
+            nesting++;
+            unsigned pos = nested.length();
+            nested.append((size32_t)0);
+            return pos;
+        }
+        virtual void endNested(size32_t sizePos) override
+        {
+            size32_t sz = nested.length()-(sizePos + sizeof(size32_t));
+            nested.writeDirect(sizePos, sizeof(sz), &sz);
+            nesting--;
+            if (!nesting)
             {
-                while (rhsState.pos < rhsState.arr->numRows)
-                {
-                    Row &row = rhsState.arr->rows[rhsState.pos++];
-                    if (row.rhs)
-                    {
-                        fpos = row.fpos;
-                        return row.rhs;
-                    }
-                }
+                iFileIOStream->write(nested.length(), nested.toByteArray());
+                nested.clear();
             }
-            rhsState.arr++;
-            rhsState.pos = 0;
         }
-        fpos = 0;
-        return nullptr;
+    public:
+        CJoinGroupRowWriter(IFile *iFile, IOutputRowSerializer *_serializer, bool append, bool compress) : serializer(_serializer)
+        {
+            if (compress)
+            {
+                iFileIO.setown(createCompressedFileWriter(iFile, 0, append, false, nullptr, COMPRESS_METHOD_LZ4));
+                iFileIOStream.setown(createIOStream(iFileIO));
+            }
+            else
+            {
+                iFileIO.setown(iFile->open(append ? IFOreadwrite : IFOcreate));
+                iFileIOStream.setown(createBufferedIOStream(iFileIO));
+            }
+            if (append)
+                iFileIOStream->seek(0, IFSend);
+        }
+        ~CJoinGroupRowWriter()
+        {
+            iFileIOStream.clear(); // flushes
+        }
+        void write(const void *row, offset_t fpos)
+        {
+            serializer->serialize(*this, (const byte *)row);
+            iFileIOStream->write(sizeof(fpos), &fpos);
+        }
+        offset_t queryPosition() const
+        {
+            return iFileIOStream->tell();
+        }
+        void addMarker()
+        {
+            CJoinGroup::SpillMarker marker;
+            iFileIOStream->write(sizeof(marker), &marker);
+        }
+        void updateMarker(offset_t pos, const CJoinGroup::SpillMarker &marker)
+        {
+            iFileIOStream->flush();
+            iFileIO->write(pos, sizeof(marker), &marker);
```

Comment:
it's not, so it is not using compressed writer in this case.
i.e. if it's dealing with an incomplete group (after looking for any complete groups to spill 1st), then it will use an uncompressed writer here.